### ISSUE TITLE
Collection notification not working when delete objects and add objects in same transaction

### ISSUE THREAD

**Issue by marcuswu0814** [realm/realm-cocoa#7027]
## Goals

Trigger collection notification.

Want to check is this except behavior or bug?

## Expected Results

The observe closure shuold receive changeset type `.update()`.

## Actual Results

In `10.5` and `5.x`, nothing receive. 

But in `4.4.1`, it receive update notification.

## Steps for others to Reproduce

Tap add button and tap update button to clone old objects and delete them, re-add them into realm.

## Code Sample

```swift
func viewDidLoad() {
  super.viewDidLoad()
  
  let realm = try! Realm()
  token = realm.objects(PersonObject.self)
    .observe { changeset in
      switch changeset {
        case let .initial(latestValue):
          print("🔆 Init \(latestValue)")
                    
        case let .update(latestValue, _, _, _):
          print("🔆 Update \(latestValue)")
                    
        case let .error(error):
          print("🔆 Error \(error)")
        }
    }
    
    addButton.rx.tap
      .subscribe(onNext: { _ in
        print("👇 Add button tap.")

        let realm = try! Realm()
                
        try! realm.write {
        let newObject = PersonObject()
                    
        let string = UUID().uuidString
        newObject.id = string
        newObject.name = string
                    
        realm.add(newObject)
      }
    })
    .disposed(by: disposeBag)
    
    updateButton.rx.tap
      .subscribe(onNext: { _ in
        print("👇👇 Update button tap.")
                
        let realm = try! Realm()
                
        let oldObjects = realm.objects(PersonObject.self)
        let newObjectsTemplate = Array(oldObjects).map { $0.toTemplate() }
        let newObjects = newObjectsTemplate.map { $0.toObject() }
                
        newObjects.forEach { $0.isBusy = true }
                
        try! realm.write {
          realm.delete(oldObjects)
          realm.add(newObjects)
        }
        
        // If doing write transaction separate to delete and add, the notification work(2 separate notification).
      })
      .disposed(by: disposeBag)
}
```

## Version of Realm and Tooling

Realm framework version: ? **10.5**

Realm Object Server version: ? None

Xcode version: ? **12.2**

iOS/OSX version: ? **14.2**

Dependency manager + version: ? **CocoaPods 1.10.0**

**Comment by DominicFrei** [comment]
Hi @marcuswu0814,

I've created an example to test what you are describing:
https://github.com/DominicFrei/Playground/tree/realm/realm-cocoa/issues/7027

Is this what you are looking for? 
If so, can you maybe provide an example of your non-working version with a little more context?

**Comment by tgoyne** [comment]
It sounds like this will be fixed by https://github.com/realm/realm-core/pull/4167 once that is included in a release.

**Comment by marcuswu0814** [comment]
Hi @DominicFrei, @tgoyne

This is [example project](https://github.com/marcuswu0814/Realm-Collection-Notification-Bug). You can checkout different branch to see my result.

I try to install realm from main branch version and it still not work, wanna check how to update `realm-core` version with CocoaPods?

**Comment by DominicFrei** [comment]
Thanks, @tgoyne!

@marcuswu0814 Can you have a look into the PR linked by Thomas? This should solve the problem. It's already merged but needs to be released.

**Comment by marcuswu0814** [comment]
The bug seems be fixed in `10.5.1` in example project.

**Comment by DominicFrei** [comment]
That's great to hear. Thanks for the feedback, @marcuswu0814
